Index: dpkt-svn/dpkt/__init__.py
===================================================================
--- dpkt-svn/dpkt/__init__.py	(revision 54)
+++ dpkt-svn/dpkt/__init__.py	(working copy)
@@ -21,6 +21,12 @@
 import dns
 import dtp
 import esp
+
+import rtap
+import llc
+import prism
+import ieee80211
+
 import ethernet
 import gre
 import gzip
Index: dpkt-svn/dpkt/http.py
===================================================================
--- dpkt-svn/dpkt/http.py	(revision 54)
+++ dpkt-svn/dpkt/http.py	(working copy)
@@ -16,8 +16,10 @@
         if not line:
             break
         l = line.split(None, 1)
+        # Sometimes we can split TCP packets here too.
         if not l[0].endswith(':'):
-            raise dpkt.UnpackError('invalid header: %r' % line)
+            raise dpkt.NeedData('short header line: %r' % line)
+            #raise dpkt.UnpackError('short header line: %r' % line)
         k = l[0][:-1].lower()
         d[k] = len(l) != 1 and l[1] or ''
     return d
Index: dpkt-svn/tests/test-perf2.py
===================================================================
--- dpkt-svn/tests/test-perf2.py	(revision 54)
+++ dpkt-svn/tests/test-perf2.py	(working copy)
@@ -1,5 +1,8 @@
 #!/usr/bin/env python
 
+#import psyco
+#psyco.full()
+
 import dpkt
 import time, unittest
 
